### PR TITLE
Sparse fake tensor support

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #82172

Add support for sparse fake tensors.

- The testing strategy is to run a fake tensor cross ref test on `test_sparse.py`. This is necessary because OpInfo sparse coverage is completely nonexistent. We could have tried to turn on cross ref testing globally for all files, but that would be very time consuming and the tests I'm interested in are mostly in this file. There are some exclusions in testing for things that don't work.
- I make fake tensor converter raise a UnsupportedFakeTensorException if the meta converter fails to do a conversion (which can happen in a relatively large number of situations).
- I relax fake tensor invariants so that you can make a fake tensor from a meta tensor. This is useful because in the cross ref test sometimes we operate on meta tensors.
- Fake tensor wrapping is improved to handle the case when a function doesn't return any tensors
- Meta converter is taught how to convert sparse tensors to meta

There's still a little more cleanup that needs to be done, but this is good for review.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

**Comment by ezyang** [comment]
cc @eellison my plan is to carve off most of the pieces here into standalone PRs

File: torch/_subclasses/fake_tensor.py [hunk 1]
```diff
@@ -613,24 +629,39 @@ def check_non_fake_tensor(x):
...
-                    r = run_fallback_kernel(func, args, kwargs, not_implemented_error)
+                    return run_fallback_kernel(
+                        self, func, args, kwargs, not_implemented_error
+                    )
 
             # TODO: handle non-kwarg devices
             assert func not in _device_not_kwarg_ops, f"NYI: {func}"
 
+            # Lazily initialized, in case there are no tensor returns
+            common_device = None
```
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:640]: fwiw, I stole this part of your code to get https://github.com/pytorch/pytorch/pull/82449 to pass hehe (otherwise aten.maximum_value or something along those lines with no Tensor inputs would fail)

File: test/test_sparse.py [hunk 2]
```diff
@@ -40,7 +41,53 @@
...
+                    fake_r = func(*fake_args, **fake_kwargs)
+            except UnsupportedFakeTensorException:
+                pass
+
+        r = func(*args, **kwargs)
+        return r
+
+class TestSparseBase(TestCase):
+    def run(self, result=None):
+        if TEST_WITH_CROSSREF:
```
Comment (eellison) [inline test/test_sparse.py:84]: there are separate `TEST_WITH_CROSSREF` ci runs, is that correct ?

File: torch/_subclasses/fake_tensor.py [hunk 3]
```diff
@@ -148,11 +149,21 @@ def from_meta_and_device(self, fake_mode, t, device):
         self.set_tensor_memo(t, out)
         return out
 
+    # There are two ways to call this.  First, you can have manually constructed
```
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:152]: thanks for comment !

File: torch/_subclasses/fake_tensor.py [hunk 4]
```diff
@@ -324,25 +341,28 @@ def __new__(cls, fake_mode, elem, device):
...
             device = torch.device(f"cuda:{torch.cuda.current_device()}")
-        assert device.type != "meta"
         self.fake_device = device
         self.fake_mode = fake_mode
         self.has_sym_ints = symbolic_shapes.has_symbolic_sizes_strides(elem)
 
     @staticmethod
     def from_tensor(t, fake_mode):
         existing_device = t.device
+        # TODO: this should use meta converter
```
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:358]: 👍

File: torch/_subclasses/fake_tensor.py [hunk 5]
```diff
@@ -324,25 +341,28 @@ def __new__(cls, fake_mode, elem, device):
...
         return FakeTensor(fake_mode, t.to(device="meta"), existing_device)
 
     # TODO: resolve error in default __repr__
     def __repr__(self):
-        return f"FakeTensor({self.fake_device}, {self.size()}, {self.dtype})"
+        with in_kernel_invocation_manager(self.fake_mode):
```
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:363]: nice!

File: torch/_subclasses/fake_tensor.py [hunk 6]
```diff
@@ -381,6 +401,8 @@ def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
                 return torch.device("meta")
             else:
                 return args[0].fake_device
+        elif func == torch.ops.aten.stride.default:
```
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:404]: just curious, why was this necessary/what does this accomplish ?

File: test/test_sparse.py [hunk 7]
```diff
@@ -40,7 +41,53 @@
...
+                    return f(t)
+                else:
+                    return t
+            return go
+
+        # empty_like excluded for now due to sparse complex
+        # aten._to_dense.default this one is getting called with csc
+        if (
+            func not in [
+                torch.ops.aten.lift_fresh.default,
```
Comment (eellison) [inline test/test_sparse.py:60]: reason for lift_fresh exclusion ?

File: torch/_subclasses/fake_tensor.py [hunk 8]
```diff
@@ -657,24 +690,29 @@ def to_real_tensor(e):
...
-    # to fake tensors later
-    meta_converter = MetaConverter()
-
     def map_out(e):
-        return inp_impls.get(id(e), meta_converter(e))
+        if isinstance(e, torch.Tensor):
+            if id(e) in inp_impls:
+                return inp_impls[id(e)]
+            else:
+                return fake_mode.fake_tensor_converter(fake_mode, e)
```
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:713]: Hmm, why is this change necessary ? I guess this works now because we're using the actual devices, not cpu fallback.

File: torch/_subclasses/meta_utils.py [hunk 9]
```diff
@@ -94,7 +94,10 @@ def set_tensor_memo(self, t, v):
         # hold a weak ref to self, otherwise it will be kept alive
         # by the del_ten closure
         self_weak_ref = weakref.ref(self)
-        weak_st = StorageWeakRef(t.storage())
+        if t.is_sparse:
```
Comment (eellison) [inline torch/_subclasses/meta_utils.py:97]: Is it worth holding onto either the values or indices of the sparse tensor storage ?

File: torch/_subclasses/fake_tensor.py [hunk 10]
```diff
@@ -643,6 +674,8 @@ def run_fallback_kernel(func, args, kwargs, orig_not_implemented_exception):
         def to_real_tensor(e):
             if isinstance(e, FakeTensor):
                 out = torch.zeros_like(e, device=e.fake_device)
```
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:676]: so I guess `zeros_like` doesn't propagate the `coalesced`-ness of the input?

File: torch/_subclasses/fake_tensor.py [hunk 11]
```diff
@@ -324,25 +341,28 @@ def __new__(cls, fake_mode, elem, device):
         )
 
     def __init__(self, fake_mode, elem, device: Union[torch.device, str]):
-        # elem does not need to be recorded, because FakeTensor *is a* elem
-        assert elem.device.type == "meta", elem
+        assert elem.device.type == "meta", elem.device.type
         device = device if isinstance(device, torch.device) else torch.device(device)
+        # NB: it is fine, if a little confusing, for device to be meta
```
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:346]: Hmm, what do you think about making allowing this a property of the FakeTensorMode ? Ive had a few instances where these assertions have caught bugs.

**Review by eellison — COMMENTED** [review]
Cool!! A few comments for my clarification but looks good

File: test/test_sparse.py (hunk 2 above)
Comment (ezyang) [reply test/test_sparse.py:84]: Yes. CPU only right now.

File: torch/_subclasses/fake_tensor.py (hunk 6 above)
Comment (ezyang) [reply torch/_subclasses/fake_tensor.py:404]: TBH, I kind of have forgotten. I think it was a symbolic shapes / fake tensor interaction, maybe.

File: test/test_sparse.py (hunk 7 above)
Comment (ezyang) [reply test/test_sparse.py:60]: also don't remember; but lift_fresh is weird and it's not really that important to test here haha

File: torch/_subclasses/fake_tensor.py (hunk 8 above)
Comment (ezyang) [reply torch/_subclasses/fake_tensor.py:713]: So, the problem I was having is that the meta converter would sometimes fail to convert the output, resulting in a non meta tensor, which would then fail later. It seemed cleaner to go straight to fake tensor here.

File: torch/_subclasses/meta_utils.py (hunk 9 above)
Comment (ezyang) [reply torch/_subclasses/meta_utils.py:97]: So, the sparse tensors are not supposed to alias with each other internally, so hypothetically this is not necessary. However, it is possible to get an aliasing reference to values and is often used by users to do an inplace modification to a sparse tensor. This needs to be worked out how to do correctly...

File: torch/_subclasses/fake_tensor.py (hunk 10 above)
Comment (ezyang) [reply torch/_subclasses/fake_tensor.py:676]: No, because you always end up with nnz=0 for zeros :)

File: torch/_subclasses/fake_tensor.py (hunk 11 above)
Comment (ezyang) [reply torch/_subclasses/fake_tensor.py:346]: Yeah, I'm cool with that. Can add it.

File: torch/_subclasses/fake_tensor.py (hunk 6 above)
Comment (ezyang) [reply torch/_subclasses/fake_tensor.py:404]: I removed it and remembered. I'll push a comment

File: torch/_subclasses/fake_tensor.py [hunk 12]
```diff
@@ -466,9 +501,10 @@ def merge_devices(t):
 
 
 class FakeTensorMode(TorchDispatchMode):
-    def __init__(self, allow_fallback_kernels=True):
+    def __init__(self, *, allow_fallback_kernels=True, allow_meta=False):
```
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:504]: nice... is it generally good practice to have keyword-only args ? just allows more flexibility with respect to changes in the future basically ?

File: torch/_subclasses/fake_tensor.py [hunk 13]
```diff
@@ -658,24 +705,29 @@ def to_real_tensor(e):
 
         for e in tree_flatten((args, kwargs))[0]:
             if isinstance(e, torch.Tensor):
-                storages.add(e.storage()._cdata)
+                if not e.is_sparse:
+                    storages.add(e.storage()._cdata)
```
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:709]: I would maybe add a `# TODO: sparse storage aliasing` here

File: torch/_subclasses/meta_utils.py (hunk 9 above)
Comment (eellison) [reply torch/_subclasses/meta_utils.py:97]: Maybe add a `# TODO: sparse aliasing` here ?

**Review by eellison — APPROVED** [review]
Nice!!!

File: torch/_subclasses/fake_tensor.py (hunk 12 above)
Comment (ezyang) [reply torch/_subclasses/fake_tensor.py:504]: I prefer kwargs to avoid boolean blindness; if I write `FakeTensorMode(True)` what does it mean? Can't tell without looking at the call site.

File: torch/_subclasses/fake_tensor.py (hunk 13 above)
Comment (ezyang) [reply torch/_subclasses/fake_tensor.py:709]: ok will do in follow up

**Comment by eellison** [comment]
yea you'll need to rebase bc i copy-pasta'd your code here early haha
